### ISSUE TITLE
Define the origin w.r.t. the pixel corner or center

### ISSUE THREAD

**Issue by constantinpape** [ome/ngff#89]
The image specification (`multiscales`) must define the origin, either w.r.t. pixel center or corner.

@lassoan summarized this well in https://github.com/ome/ngff/pull/85#issuecomment-1026030204:
> In 2D applications using one of the pixel corners as origin is probably more common, but in 3D imaging voxel coordinate system origin is typically the center. In VTK, ITK, and all applications based on these toolkits origin is in the pixel center. In all 3D image file formats that I know of (nrrd, nifti, metaio, dicom) origin is in the pixel center, too.
> NGFF standard must specify this (if it hasn't been specified it already).
> If any software that uses different coordinate system convention internally then it can convert to/from the standard coordinate system when reading/writing images.

So I think we should use the pixel center as origin here as well. Anyone wants to make a small PR to add this, @jbms @d-v-b @lassoan @thewtex? Otherwise I can give this a shot once #85 is merged.

**Comment by bogovicj** [comment]
> So I think we should use the pixel center as origin here as well.

Yes, agreed.

**Comment by lassoan** [comment]
For reference, here is how ITK defines image geometry (origin, spacing, directions): https://itk.org/ITKSoftwareGuide/html/Book1/ITKSoftwareGuide-Book1ch4.html#x45-540004.1.4

**Comment by constantinpape** [comment]
Thanks for providing the link @lassoan! I read this now and in light of yesterdays discussion I think that we should not add it to 0.4, but instead tackle this in 0.5, given that @bogovicj will also make the definition of spaces and data-spaces much more explicit than it currently is. With this the origin definition will make much more sense.

If anyone thinks we need a sentence on it right now, please go ahead and propose changes in a PR; I am not opposed to adding it for v0.4 (or to make a v0.4.1 with just that change), but I can't find a good place to put this in right now without the more explicit space definition.

**Comment by xulman** [comment]
> So I think we should use the pixel center as origin here as well.
> 
> Yes, agreed.

Besides, there's *center* while 4, 8, ? corners... and one would have to specify which corner

**Comment by jbms** [comment]
The corner would surely be the start (i.e. top left) along each dimension, so I don't think there is an issue there.

Though I suppose you could in theory allow a choice of center vs "corner" (i.e. boundary) independently for each dimension.

I have a vague intuition that if the data will be displayed using interpolation then center is a more natural choice, while if the data will be displayed without interpolation (i.e. "pixelated") then corner may be a more natural choice.

For example:

Suppose we have a 1-d array of size 10 at a resolution of 4nm.

If we say integer coordinates are at the corner, then our data corresponds to the continuous physical range [0nm, 40nm] assuming we don't apply any translation.

If we instead had an array of size 5 at a resolution of 8nm, then our data would still correspond to the continuous physical range [0nm, 40nm].

If we say integer coordinates are at the center, then our data corresponds to the continuous physical range [-2nm, 38nm] if we don't apply any translation.  If we instead had an array of size 5 at a resolution of 8nm, then we would instead get by default a continuous physical range [-4nm, 36nm].

On the other hand, if we are always interpolating, and will exclude the outer half of the begin and end pixels, then our array of 10 4nm pixels would correspond to the physical range [0, 36nm] and our array of 5 8nm pixels would correspond to the physical range [0, 32nm] which is maybe a bit more intuitive.

When I made the choice in Neuroglancer to use corner rather than center it was more a natural result of how the transforms were implemented rather than a conscious decision, and then changing it later was not an option for backwards compatibility reasons, so I'm curious what others think about this.

**Comment by thewtex** [comment]
Top left is does correspond to how most visualization tools orient their rendering. However, bottom left is how most processing tools orient themselves. By using the center, we avoid these issues.

We want to support processing and rendering in these non-"pixelated" methods, e.g. @tpietzsch demo'ed rendering a splat-type rendering with BigDataViewer.  Pixelated, interpolated-pixelated, non-pixelated, processing can all be supported when the  transform applies to the pixel center. Considering all use cases, complexities and dependencies on the size of a pixel and which corner is taken to be the start are removed.

**Comment by lassoan** [comment]
In radiology (and in 3D medical imaging software libraries and applications I know), the debates around pixel corner/center took place in the early 2000s and the community standardized on pixel center. The decision was not contested later and in general everyone is happy with it.

Image is a continuous signal that can be reconstructed flawlessly from the discrete samples stored in the voxel values (as long as the Nyquist sampling criterion was respected). Therefore pixelated display is unnecessary, and it can be also simply considered incorrect, because we know that the original signal can be reconstructed using a low-pass filter yet we construct a signal using a zero-order hold. Some people switch to pixelated display because they want to see the voxel boundaries, but this goal can be achieved much better by overlaying a grid (the grid always shows voxel boundaries clearly, regardless of brightness/contrast settings and intensity difference between neighbor voxels).

> then changing it later was not an option for backwards compatibility reasons, so I'm curious what others think about this.

I can tell about a similar issue that we have in 3D Slicer. At the time when the application was designed, image axis orientation in radiology software was still not standardized. Slicer chose RAS, while over the years the rest of the radiology world ended up using LPS. Slicer kept using RAS for backward compatibility reasons. As time went on, we encountered more and more issues due to this inconsistency, but as more and more features were added, more data was generated, the larger the community grew, switching to the standard convention just got so much more complex and bigger task that it has remained out of our reach. With several years of careful work we managed to switch to standard LPS in all the files, but internally we still use RAS. This is a source of lots of complications, potential errors, and user confusion. It would have been much better to switch many years ago, when everything was still smaller and simpler.

This is of course just one example, but it illustrates how you can get into a tough spot if your software diverges from common conventions in your field. The change to using pixel center when representing 3D images with a 3D array may look hard and/or unnecessary, but in the long term it will be just harder and it may turn out that the change is practically unavoidable (because you just need to spend too much time debugging and fixing errors related to half-voxel offset errors and explaining to users and developers why they need to add half voxel offsets here and there when they use or write plugins for your software).

**Comment by thewtex** [comment]
@jbms yes, that's right -- we want to specify `O` vs `A`. While we avoid any confusion with the orientation of the coordinate, we also support processing as @lassoan pointed out.

**Comment by bogovicj** [comment]
[Here's a blurb I wrote re this definition that I hope will be in the next version](https://github.com/bogovicj/ngff/blob/coord-transforms/latest/index.bs#L283-L292):


> ### Coordinate convention
>
>**The pixel/voxel center is the origin of the continuous coordinate system.**
>
> It is vital to consistently define relationship between the discrete / arrray and continuous / interpolated
> coordinate systems. A pixel / voxel is the continuous region that corresponds to a single sample
> in the discrete array, i.e., the area corresponding to nearest-neighbor (NN) interpolation of that sample.
> The center of a 2d pixel corresponding to the origin `(0,0)` in the discrete array is the origin of the continuous space
> `(0.0, 0.0)` (when the transformation is the identity). The continuous rectangle of the pixel is given by the
> half-open interval `[-0.5, 0.5) x [-0.5, 0.5)`. See chapter 4 and figure 4.1 of the ITK Software Guide [[itk]].

**Comment by dstansby** [comment]
What's the status of this discussion? Is it waiting on someone to open a pull request with a proposal?

**Comment by d-v-b** [comment]
Regarding the status of the discussion, going by emoji and vibes my read is that most participants are in the "pixel center" camp, but that doesn't mean everyone agrees. 

I think the "pixel center or pixel corner" question is basically about how to draw little squares on a screen; this is an important concern, but I think it's out of scope for a file format specification for data sampled in physical coordinates. Zarr arrays don't contain pixels, they just contain values. We use metadata to assign those values to positions in space, i.e. we assign each value to a _point_, (not a square / cube / hypercube). The little squares only enter the story when someone decides to display those values by resampling onto a display coordinate space after using nearest neighbor interpolation, which is just one of several possible interpolation schemes.

As for the future of the spec, I believe John will cover these points better than I can in his forthcoming transformations RFC.

**Comment by lassoan** [comment]
> I think the "pixel center or pixel corner" question is basically about how to draw little squares on a screen; this is an important concern, but I think it's out of scope for a file format specification for data sampled in physical coordinates.

I agree that the file format does not need to talk about pixels, as the data array just stores signal samples at physical locations. It is up to the visualization software to specify display coordinate system and render pixels.

However, based on the discussion in this thread, it may worth adding a note to the file format specification that explains the difference between physical and display coordinate systems.

**Comment by jbms** [comment]
The pixel origin matters for mapping between the discrete coordinate space of a zarr array and any continuous coordinate space.  In particular, when defining a multiscale, the relative transformations that need to be specified depend on the pixel origin.

**Comment by jbms** [comment]
Also, pixel origin is the center is anyway the defacto standard since that is what existing ome-zarr software and multiscale datasets use.  That should be described in the spec.

**Comment by dstansby** [comment]
To get my head around this I did some ASCII art with a toy example.

If the multiscales axes unit is meters, and the coordinate transform scale for this level is 0.2, then an array with four items maps to a physical coordinate system like this:

```
0       1       2       3       4
|       |       |       |       |
---------------------------------
|       |       |       |       |
0.0m    0.2m    0.4m    0.6m    0.8m
```

The transform is
```json
[
    {
        "type": "scale",
        "scale": [0.2]
    }
]
```

This is all the spec is for, the transform between array indices and a continous space. No pixels here!

I think there are two common issues with how an array like this is visualised, and how lower resolution multiscales data are calculated from it. These two issues cancel each other out so they don't look like issues, but in reality they are.

## Visualisation

For the above array, it is often visualised (I believe this is how neuroglancer currently works?) as:
```
0       1       2       3       4
|0000000|1111111|2222222|3333333|...
---------------------------------
|       |       |       |       |
0.0m    0.2m    0.4m    0.6m    0.8m
```

But this is doesn't seem like a very sensible way to visualise - physical coordiante (e.g.) 0.15m is closer to array value `1`, but is being rendered as array value `0`. It is worth stressing at this point that the values in the array are *not* pixel values to be inerpreted as being defined over an imaginary pixel or higher-dimensional voxel, they are values defined at a **single point** in a physical space.

As such, it makes more sense to use nearest neighbour interpolation:

```
0       1       2       3       4
0000|1111111|2222222|3333333|4444...
---------------------------------
|       |       |       |       |
0.0m    0.2m    0.4m    0.6m    0.8m
```

## Creating downsampled data
To create a lower resolution level multiscale pyramid, we can bin this data by a factor of two. That means take every two values, and average them. Where should these new values lie in the physical coordinate space? It seems sensible to put them halfway between the two values we averaged:

```
    0               1
|   |   |   |   |   |   |   |   |
---------------------------------
|       |       |       |       |
0.0m    0.2m    0.4m    0.6m    0.8m
```

The correct transform for this level is then:

```json
[
    {
        "type": "translation",
        "translation": [-0.5]
    },
    {
        "type": "scale",
        "scale": [0.4]
    }
]
```
But the translation is easy to miss (certainly, software I have written currently misses the translation and I only noticed when writing this comment 😱). Without the translation we would have:

```
0               1   
|   |   |   |   |   |   |   |   |
---------------------------------
|       |       |       |       |
0.0m    0.2m    0.4m    0.6m    0.8m
```

If you display this array using the first option in Visualisation above, the pixels overlap with the pixels they have been binned from, so two wrongs have consipred to make something that looks right!

I am guessing this is the thinking the radiology community went through as mentioned in https://github.com/ome/ngff/issues/89#issuecomment-1029055360. I used to work in astronomy, where the same issue was also present in the past until we did a drive to educate folks on how array values mapped to pixels (ie, your array value is defined at the centre of a pixel, not the corner).

Sorry for the long post, but I hope it explains stuff for others as well as me? The conclusions I'm getting from this and the conversation above are:

- The spec does not need to mention anything about pixels
- ... but as a community we need to educate folks to get their transforms right when creating multiscale data, and check and potentially correct our own software

**Comment by d-v-b** [comment]
>     {
>        "type": "translation",
>       "translation": [-0.5]
>     }

I think the sign is wrong here -- if your original coordinates were `[0, 1, 2, 3, ...]`, then downsampling by 2x windowed averaging will yield coordinates `[0.5, 2.5, ...]`.

**Comment by dstansby** [comment]
Thanks for catching, I think fixed now!

**Comment by almarklein** [comment]
Hi!

Let me say upfront that I wholeheartedly agree with the pixel-center convention. I am the main author of [PyGfx](https://github.com/pygfx/pygfx), and our `Image` and `Volume` objects also follow this convention. Bluntly said, you want to point at the pixels, not at the boundary between them :)

I am currently working on a tool / viewer that consumes Zarr images, and I am trying to find out how to best deal with this situation, so that my tool handles both old and new data correctly 😅  Let me share my thoughts ...

### pixel-corner vs pixel-center

One thing I noticed is that ome-zarr-py's `write_image` does not write translations to the metadata. IIUC the resulting data should be interpreted like this (using @dstansby's art as a starting point):

```
      0.0m    0.2m    0.4m    0.6m    0.8m
      |       |       |       |       |
-----------------------------------------

      |       |       |       |       |   level 0, scale 0.2
      
      |               |               |   level 1, scale 0.4
      
      |                               |   level 2, scale 0.8
```

This implies it follows the pixel-corner convention, resulting in zero translation.

Let's transform this drawing to the pixel-center convention, drawing the same image, but with circles for the pixel centers:

```
      0.0m    0.2m    0.4m    0.6m    0.8m
      |       |       |       |       |
-----------------------------------------

      |   o   |   o   |   o   |   o   |   level 0, scale 0.2, translation 0.1
      
      |       o       |       o       |   level 1, scale 0.4, translation 0.2
      
      |               o               |   level 2, scale 0.8, translation 0.4
```

Now you do have translations, which are easy to calculate: ``translation = scale / 2``.

However, the pixel at level 0 is not lined up with the origin. Let's fix that too:

```
      0.0m    0.2m    0.4m    0.6m    0.8m
      |       |       |       |       |
-----------------------------------------

  |   o   |   o   |   o   |   o   |       level 0, scale 0.2, translation 0
      
  |       o       |       o       |       level 1, scale 0.4, translation 0.1
      
  |               o               |       level 2, scale 0.8, translation 0.3
  ```

The translation now is ``translation = scale / 2 - scale_at_level_0 / 2``.

### Best practices?

The approach that I'll probably take is:

* Use pixel-center convention internally (for processing and visualization).
* If translations are given, assume the pixel-center convention. 
* If translations are omitted, calculate translations automatically, mapping to either the 2nd or 3d image *edit: not sure which one, tbh)*. 

I think it would be helpful if the spec recommended an approach on how to interpret the absence of translations. 

Happy to hear other people's thoughts on this!

**Comment by jbms** [comment]
The OME zarr spec doesn't distinguish between "unspecified" translation and zero translation, and we've agreed that transforms in OME zarr are based on the convention that integer coordinates correspond to the center of the pixel.  I'd strongly suggest that you follow the spec and not add any offsets to the translation based on heuristics.  Instead, the writers need to be fixed to correctly specify the correct translation for each scale.

**Comment by almarklein** [comment]
I'd love that, but as it is now, some writers adopt the pixel-corner convention and some the pixel-center convention. Even the current `ome-zarr-py`s writer produces a zarr-file that adopts the pixel-corner convention. 

And it *does* matter, and IMO should be a concern of the spec, because if a viewer assumes a different convention than the writer, the different layers will not be aligned and the layer zero will have an unintended half-pixel offset. When users then make annotations based on this image data, these annotations may be wrong.

Maybe it's worth taking a small step back, and explain what I think that the different conventions actually mean ...

### Downsampling
You can downsample data in many different ways. You could, for instance, smooth the data and then take every other sample `a[::2]`:
```
             _______ _______ _______ _______
            |___o___|___o___|___o___|___o___|   level 0
         _______________ _______________ ___
        |_______o_______|_______o_______|___    level 1
 _______________________________ ___________
|_______________o_______________|___________    level 2
```

You can see how the pixels line up, and the corners are offset w.r.t. each-other. So the offsets become:
* pixel-center convention: `0`
* pixel-corner convention: `- scale / 2`

However, it seems much more common, to average each two subsequent pixels, i.e. to take the new samples in between two pixels. This is also what e.g. `skimage.transform.resize` and friend do:
```
       _______ _______ _______ _______
      |___o___|___o___|___o___|___o___|   level 0
       _______________ _______________
      |_______o_______|_______o_______|   level 1
       _______________________________
      |_______________o_______________|   level 2
```
Now the corners line up, but the pixels do not. So the offsets become:
* pixel-center convention: `scale/2`
* pixel-corner convention: `0`

### Interpretation by a viewer

So, if I encounter a zarr file that does not write translations, it could be because it uses the top downsampling method, and pixel centers indeed already line up. But it's way more likely that it's the bottom one, and the *corners* already line up, i.e. it implicitly adopts the pixel-corner convention.

So if my tool adopts the pixel-center convention, and it encounters data without translations, it's pretty safe to assume that it uses pixel-corner conventions, and I am going to convert it to pixel-center convention by using offset `scale/2`. And again, data written with ome-zarr-py (with minimal args) falls into this category.

Another way to say this: if a writer uses the pixel-center convention, it *must* write translations, unless it uses a more special downsampling method as shown at the top of this post.

I'd love it if there was a way to explicitly tell whether data was written with the pixel-corner or pixel-center convention, but form what I can see the most reliable way is to check whether translations are present 🤷

**Comment by d-v-b** [comment]
I wouldn't frame this as tools adopting one convention or another, when it's far more likely that some tools simply fail to write correct coordinate transformations. This of course can happen no matter how clear the spec is. 

Your example conflates the procedure used to generate downsampled intensities (averaging vs subsampling) with the choice of the output coordinate grid (subset of the original grid vs a new coordinate grid). I am pretty sure that nearly every tool that prepares multiscale data for visualization purposes targets the same output coordinate grid -- namely one that converges to the center of the source grid. But as a tool writer I would not assume anything here, and put the burden of generate correct data on the producer.

**Comment by d-v-b** [comment]
If users of your tool consume a lot of data that wasn't created with the correct metadata, and it's generally accepted what the correct metadata _should_ be, then your tool could add an option to "apply X correction to input OME zarr metadata". This could e.g. transform the input metadata document _before_ its interpreted as defining assets in a scene. Users could be notified that their OME zarr metadata wouldn't need the correction step if the translation transformation were specified, etc.

**Comment by will-moore** [comment]
Just opened at PR at https://github.com/ome/ome-zarr-py/pull/590 to fix this

**Comment by almarklein** [comment]
> Your example conflates the procedure used to generate downsampled intensities (averaging vs subsampling) with the choice of the output coordinate grid (subset of the original grid vs a new coordinate grid).

I agree that there are indeed two questions:
* how level 0 is mapped into world space (micrometers).
* how the different levels are aligned on top of each-other.

But both are answered the same way: metadata translations.

The chosen downsampling procedure only affects the latter. Nevertheless, for the common downsampling procedures, you need offsets to align the pixel centers. And not providing offsets suggests it intends to aligns by the corners, so I'd expect level 0 to be aligned by the corner also.

> I wouldn't frame this as tools adopting one convention or another, when it's far more likely that some tools simply fail to write correct coordinate transformations. This of course can happen no matter how clear the spec is.

Well, they fail to write "correct" transformations because they've (consciously or implicitly) adopted a different coordinate frame (pixel-corner vs pixel-center). And this happens because the spec up until 0.5 was not clear about this, so you cannot really blame it on the tools. (I don't mean this to criticize the spec, I find it quite beautiful in most regards). 

----

As a result we are now in a transitional phase with both conventions being widely present, and we should think how to best move forwards from here.

* Document how writing-tools should write translations.
* Think of a way for reading-tools to distinguish between the two conventions. My best guess is checking whether translations are present, but love to hear of a better way.

**Comment by d-v-b** [comment]
> Well, they fail to write "correct" transformations because they've (consciously or implicitly) adopted a different coordinate frame (pixel-corner vs pixel-center).

I would check this with the source of the data you are working with. You mentioned ome-zarr-py -- in that case, I think the omission of translation transform metadata is a [bug](https://github.com/ome/ome-zarr-py/issues/403), rather than adherence to an alternative convention.

**Comment by almarklein** [comment]
Does 0.6 require translations to be present? Maybe that would help, because it forces writers to think about what the offsets should be.

I suspect that writers that omit the translations think "position the layers as one might reasonably expect", rather than "use zero offsets".

**Comment by d-v-b** [comment]
That has been requested before and I agree that it's a good idea.

**Comment by thewtex** [comment]
Hi folks, please review:

https://ngff.openmicroscopy.org/rfc/5/index.html#coordinate-convention

and whether we should add any of the suggested improvements, e.g. diagrams, to the pixel center description.

> require translations to be present

+1

CC @jo-mueller

**Comment by thewtex** [comment]
> require translations to be present

@m-albert did a good write-up on what to include in the spec here:  https://github.com/ome/ngff/issues/494

Folks working on RFC-5 also discussed this at the OME-Zarr hackathon at EMBL last week and concluded that it will be included in OME-Zarr version 0.6, guided by @jo-mueller 's leadership.
